### DISCUSSION TITLE
Como fazer transações atômicas com o Firebase Firestore

### DISCUSSION THREAD

**Discussion by EuHigorBarbosa** [Flutterando/forum#104] [Perguntas e Respostas] [answered]
Olá pessoal. Tudo bem? 
O assunto que trago é sobre dúvidas como como proceder para tratar possíveis erros em funções que se comunicam com o banco de dados. 

Por exemplo...Eu tenho um programa que utiliza o Firebase como Banco de dados. A autenticação é feita pelo Firebase Auth. 
Cada usuário tem um nome, email e senha entre outras coisas. 

Para cada usuário autenticado no FirebaseAuth existe um documento numa coleção USER no Firebase Firestore. 
E se o user for um terapeuta é criado também um documento na coleção TERAPEUTA. 

Então. se um usuário modifica o seu nome haverá uma modificação na autenticação (FirebaseAuth) , uma modificação na coleção USER e na coleção TERAPEUTA caso o usuário seja um terapeuta. 

E aí que vem a questão, vamos supor que NÃO haja erros na modificação de nome no Firebase Auth e na coleção USER do firestore mas que HAJA um erro quando da modificação do nome da coleção TERAPEUTA. 

Daí haverá uma situação com informações atualizadas para FirebaseAuth e USER, e desatualizadas para TERAPEUTAS. 

Alguém me indica uma sugestão de como tratar esse erro?  

Abaixo, como exemplo eu deixo a função que estou utilizando 
![image](https://user-images.githubusercontent.com/87345837/208999828-4c44be46-445f-4e72-8e0d-75e92f568207.png)

```
@override
  Future<void> updateDisplayName({required String displayName}) async {
    var hasInternetConnection = await internetConnectionService.checkInternetConnection();
    if (hasInternetConnection){
      try {
        //? 1. Update do valor do DisplayName no FireBase Auth
        await FirebaseAuth.instance.currentUser!.updateDisplayName(displayName);

        //? 2. Conversão para entidade de usuário do domínio do programa
        _currentUser = _toMyHealingUser(user: FirebaseAuth.instance.currentUser!, 
          name: FirebaseAuth.instance.currentUser!.displayName);
        
        //? 3. salvar o usuário na collection User FireStore
        DataBaseService dataBaseService = DataBaseService();
        CustomClaimData? customClaim = await currentUserClaims;
        await dataBaseService.updateUserOnDataBaseFireStore( 
          email:_currentUser!.email,   imageURL: _currentUser!.imageURL, 
          name:_currentUser!.name,   customClaim: customClaim!
        );

        //? 4. salvar a modificação na collection Terapeutas do FireStore
        if(customClaim.ehTerapeuta){
          await dataBaseService.updateTerapeutaOnDataBaseFireStore(email: _currentUser!.email, name: _currentUser!.name);
        }
      } on FirebaseAuthException catch (authError) {
        //! Enviar para o CrashAnalitcs!!!
        String transcriptError = FirebaseExceptionsMessages.errorsMapTranscription[authError.code]??authError.message!;
        throw AuthLoginUpdateUserDisplayNameError(
            authError.stackTrace,
            authError.code,
            authError.toString(),
            transcriptError // Erro traduzido para o português...não retirar
            );
      } on Failure catch (e,stackTrace) {
        throw AuthError(
          stackTrace,
          'AuthService-updateDisplayName',
          e,
          e.errorMessage);
      } catch (e,stackTrace) {
        throw UnknowAuthError(
          stackTrace,
          'AuthService-updateDisplayName',
          e,
          'Erro desconhecido ao fazer o update do nome de usuário (' + e.toString() + ')');
      }
    } else {
      throw InternetConnectionAuthError(
          null,
          'Erro de falta de conexão com a internet',
          'Erro de falta de conexão com a internet',
          'Erro de falta de conexão com a internet ao fazer o update do nome de usuário' 
          );
    }
  }
```

**Comment by renanzdm** [comment, marked as answer]
Acredito que o que você está tentando seria algo como uma Transaction que é usado em banco de dados relacionais. Não sei se tem algo pronto para isso no firebase, poderia fazer algo manual como manter um cache na hora em que você executa a função, se não ocorrer erros blz. Se acontecer um erro você chama as funções que deram certo alterando para o dado antigo

**Comment by EuHigorBarbosa** [comment]
Renan, era exatamente disso que eu precisava e não sabia me expressar. Precisava saber sobre Transactions e sobre batches writes que são operações atômicas utilizando o firestore. Meu muito obrigado pela ajuda. Ajudou demais me dando o norte. 

Aos interessados eu li esse artigo simples e aprendi
https://medium.com/@debnathakash8/firestore-transactions-batchedwrites-with-flutter-e675c941572f
E também vi esse vídeo: https://www.youtube.com/watch?v=n4j5776d6T4
